decoder/routes.go

```diff
@@ -0,0 +1,204 @@
+		old.Waypoints != new.Waypoints
+}
+
+func saveRouteRecord(db db.DbDetails, route *Route) error {
+	oldRoute, _ := getRouteRecord(db, route.Id)
+
+	if oldRoute != nil && !hasChangesRoute(oldRoute, route) {
+		return nil
+	}
+
+	if oldRoute == nil {
+		_, err := db.GeneralDb.NamedExec(
+			`
+			INSERT INTO route (
+			  id, name, description, distance_meters, 
+			  duration_seconds, end_fort_id, end_image, 
+			  end_lat, end_lon, image, image_border_color, 
+			  reversible, start_fort_id, start_image, 
+			  start_lat, start_lon, tags, type, 
+			  updated, version, waypoints
+			)
+			VALUES
+			  (
+				:id, :name, :description, :distance_meters, 
+				:duration_seconds, :end_fort_id, 
+				:end_image, :end_lat, :end_lon, :image, 
+				:image_border_color, :reversible, 
+				:start_fort_id, :start_image, :start_lat, 
+				:start_lon, :tags, :type, :updated, 
+				:version, :waypoints
+			  )
+			`,
+			route,
+		)
+
+		if err != nil {
+			return fmt.Errorf("insert route error: %w", err)
+		}
+	} else {
+		_, err := db.GeneralDb.NamedExec(
+			`
+			UPDATE route SET
+				name = :name,
+				description = :description,
+				distance_meters = :distance_meters,
+				duration_seconds = :duration_seconds,
+				end_fort_id = :end_fort_id,
+				end_image = :end_image,
+				end_lat = :end_lat,
+				end_lon = :end_lon,
+				image = :image,
+				image_border_color = :image_border_color,
+				reversible = :reversible,
+				start_fort_id = :start_fort_id,
+				start_image = :start_image,
+				start_lat = :start_lat,
+				start_lon = :start_lon,
+				tags = :tags,
+				type = :type,
+				updated = :updated,
+				version = :version,
+				waypoints = :waypoints
+			WHERE id = :id`,
+			route,
+		)
+
+		if err != nil {
+			return fmt.Errorf("update route error %w", err)
+		}
+	}
+
+	routeCache.Set(route.Id, *route, ttlcache.DefaultTTL)
+	return nil
+}
+
+func (route *Route) updateFromSharedRouteProto(sharedRouteProto *pogo.SharedRouteProto) {
+	route.Name = sharedRouteProto.GetName()
+	route.Description = sharedRouteProto.GetDescription()
+	route.DistanceMeters = sharedRouteProto.GetRouteDistanceMeters()
+	route.DurationSeconds = sharedRouteProto.GetRouteDurationSeconds()
+	route.EndFortId = sharedRouteProto.GetEndPoi().GetAnchor().GetFortId()
+	route.EndImage = sharedRouteProto.GetEndPoi().GetImageUrl()
+	route.EndLat = sharedRouteProto.GetEndPoi().GetAnchor().GetLatDegrees()
+	route.EndLon = sharedRouteProto.GetEndPoi().GetAnchor().GetLngDegrees()
+	route.Image = sharedRouteProto.GetImage().GetImageUrl()
+	route.ImageBorderColor = sharedRouteProto.GetImage().GetBorderColorHex()
+	route.Reversible = sharedRouteProto.GetReversible()
+	route.StartFortId = sharedRouteProto.GetStartPoi().GetAnchor().GetFortId()
+	route.StartImage = sharedRouteProto.GetStartPoi().GetImageUrl()
+	route.StartLat = sharedRouteProto.GetStartPoi().GetAnchor().GetLatDegrees()
+	route.StartLon = sharedRouteProto.GetStartPoi().GetAnchor().GetLngDegrees()
```

Comment:
I don't think we actually need to preserve start/end latlon? I think these are just the latlon for the pokestop/gym and not actually part of the route?

Reply:
I just checked a route and can confirm that these lat lon are just stop/gym lat lon. We could use them to update the gym/stop table lat lon. Not sure why this should be in the table though?

Reply:
It's uses by RM to fetch routes within few.. changing that to fort id would only cause useless joins

Reply:
Hmm, we could change it to the actual start/end latlon from waypoints, instead of the start/end poi?

Reply:
Why having a useless dependency on something that buggy implemented from nia? Why should we call for a fort when we have everything in route proto?

Nobody knows what's happening with routes if start/end fort is moved ingame, maybe route disappears ;)

But route and forts are totally different calls, keep them different as well, don't merge useless information

Reply:
I think it's clear that these lat lon are just intended to be tied to the poi itself instead of a bug as you described. A route's start/end has to be tied to a poi; it has to be near it but doesn't necessarily need to be its exact start/end point; it's part of the design?

Reply:
But then you answered your question on your own? It does not necessarily to be exact start/end point. So why modifying the route and using fort lat/lon? Useless logic, keep it as it is ;)

Reply:
For ReactMap, you ideally want to actually query the start/end lat lon, instead of the lat lon of the start/end poi, which is not actually the starting/ending point?

Reply:
The routes start and end at a poi; you have to set these when you record a route - and when you follow it is the POI which triggers the start and end.
Have you followed/created a route yet?  The route does not even appear until you go to the start POI.

Reply:
Pretty sure you can see the route even away from the POI. Are you sure you need to stand on the precise point of the POI to start the route?

Reply:
> Pretty sure you can see the route even away from the POI. Are you sure you need to stand on the precise point of the POI to start the route?

When I have completed the routes the POI is indicated with a marker pointing to it, and it has a distance countdown to the POI, and when you get close enough to the POI the route lights up.  You have to be within a normal pokemon-style distance of the POI to begin it.
Similarly with the route ending, the POI is highlighted as the endpoint that you have to get close to in order to trigger route completion.

When you create routes the first thing you do is nominate the start and end POI which is used during the recording process.  Recording can start as you get to the POI, and when you reach the final POI it pops up and asks you to confirm your route is completed.

Reply:
routes are too fresh of a feature to draw the assumption that images and coordinates are always connected to a PokeStop. the way they're sent in the proto leaves them the option to make a custom start location. drawing assumptions like this and not strictly following the protos is a big mistake we've made multiple times in the past.

Reply:
I'm not sure if I'm following the discussion. I feel like we are talking about different things. Let me maybe submit a PR and explain there.

Reply:
https://github.com/UnownHash/Golbat/pull/135